spock-core/src/main/java/org/spockframework/runtime/SpecInfoBuilder.java

```diff
@@ -226,18 +226,20 @@ private void buildFixtureMethods() {
     MethodInfo cleanupSpecMethod = createMethod(Identifiers.CLEANUP_SPEC_METHOD, MethodKind.CLEANUP_SPEC);
     if (cleanupSpecMethod != null) spec.addCleanupSpecMethod(cleanupSpecMethod);
 
-    for (Method method : clazz.getDeclaredMethods()) {
-      if (method.isAnnotationPresent(Before.class)) {
-        spec.addSetupMethod(createJUnitFixtureMethod(method, MethodKind.SETUP, Before.class));
-      }
-      if (method.isAnnotationPresent(After.class)) {
-        spec.addCleanupMethod(createJUnitFixtureMethod(method, MethodKind.CLEANUP, After.class));
-      }
-      if (method.isAnnotationPresent(BeforeClass.class)) {
-        spec.addSetupSpecMethod(createJUnitFixtureMethod(method, MethodKind.SETUP_SPEC, BeforeClass.class));
-      }
-      if (method.isAnnotationPresent(AfterClass.class)) {
-        spec.addCleanupSpecMethod(createJUnitFixtureMethod(method, MethodKind.CLEANUP_SPEC, AfterClass.class));
+    if (isJunit4SupportAvailable()) {
```

Comment:
I think this could be done with a global extension in spock-junit4

Reply:
It could be, but then it is more problematic to ensure the order of before/after elements (in that class, the super class and maybe in traits) is the same as in Spock 1.3.

Reply:
I don't think that we'll really run into problems there.

Reply:
OK. If we are fine with an another breaking change that the order of the `setup/@Before` methods might change in 2.0 I can implement it in the global extension.

Reply:
Hmm, but on the other hand it will not be intuitive to (possibly) have setup method from the current class to be executed before setup method from the super class. Cleaning up can be also problematic.

Reply:
There is no problem with super specifications, only with the order at the class level.
`"beforeClass", "setupSpec", "before", "setup", "cleanup", "after", "before", "setup", "cleanup", "after", "cleanupSpec", "afterClass"` → `[setupSpec, beforeClass, setup, before, cleanup, after, setup, before, cleanup, after, cleanupSpec, afterClass]` due to the fact when global extension is applied. It would require extra modifications in core to be able to add the methods in non-default order).

Reply:
I refactored the code to extract JUnit 4-related stuff in `SpecInfoBuilder` to an inner class.

I also extracted it to an extension, but in addition to the execution order, there is also a need to duplicate 3 methods from `SpecInfoBuilder`. I'm really not convinced that avoiding using a simple (and safe) reflection to read the check the annotations is worth to break backward compatibility.

Reply:
I switched the code to global extension.